[XPS] support multiple core levels for the same element

superstar54 opened the pull request:
## workchain support multiple core levels for the same elements

for example, `1s`, `2s`, `2p` for `Al`:

![Screenshot from 2025-02-08 11-19-35](https://github.com/user-attachments/assets/4b01e171-7fc9-4406-bd55-c1f83cce2caf)



## GUI to allow user specific the atom indices to be considered.

![Screenshot from 2025-02-08 11-23-13](https://github.com/user-attachments/assets/de83ee1a-d029-420e-8ed3-966466efeb5e)

superstar54 commented:
Hi @PNOGillespie , if you have time to review this PR today, that would be great. Otherwise, would it be okay for me to merge it today so that a user can start testing this new feature?

PNOGillespie commented on src/aiida_qe_xspec/calculations/functions/xspectra/get_xps_spectra.py:
```diff
@@ -3,109 +3,100 @@
+    Calculate the core level shift and binding energy for each element.
+    Generate the final spectra using the Voigt profile.
+
+    :param core_levels: a Dict object defining the elements and their core-levels to consider
+            when producing spectrum.
```
You should also clarify here (in addition to other docstrings) what format `core_levels` needs to be presented in

PNOGillespie commented on src/aiida_qe_xspec/workflows/xps.py:
```diff
@@ -430,39 +419,26 @@ def get_builder_from_protocol(  # noqa
-                gipaw_pseudos[element] = pseudos[element]['gipaw']
+        elements_not_present = []
+        elements_present = [kind.symbol for kind in structure.kinds]
+        if core_levels is None:
+            core_levels = {}
+            for element, pseudos in core_hole_pseudos.items():
+                if element in structure.get_symbols_set():
+                    core_levels[element] = [key for key in pseudos.keys() if key != 'gipaw']
+        for label in core_levels:
+            element = label.split('_')[0]
+            if element not in elements_present:
+                elements_not_present.append(element)
```
Since there is no input validation, `get_builder_from_protocol` will just pass in the keys of `core_hole_pseudos` for that element, regardless of whether or not those correspond to valid option, which can cause the program to except when trying to compile the final spectrum.

Some sort of input validation will be needed, in addition to deciding what inputs should be accepted as valid.

PNOGillespie reviewed (changes requested):
Hi @superstar54. The feature here does seem useful, however there are some issues regarding input validation which need resolving before this will be good to go.

I've left some comments on the various changes, but the main things which need changing are:
1. There is no input validation for `core_levels` or `core_hole_pseudos`. I understand that this automatically done when loading in your `pseudo_demo` pseudopotentials - however when working with other pseudopotentials, the lack of validation can result in the `WorkChain` excepting only at the end of a run due to simple spelling/formatting errors.
2. Docstrings (at least for the workflows) should be updated to include these new parameters, in addition to something showing how they need to be formatted.

I will have another look at this once these two points have been addressed.

superstar54 commented:
Hi @PNOGillespie , thanks for the review. I have implemented suggestions.

I have added the following validation:
- keys of `core_levels` should be a subset of the structure elements and `core_hole_pseudos`
- `atom_indices` should be in the range of the number of atoms in the structure
- all the elements corresponding to the atom indices should be in the `core_hole_pseudos`

PNOGillespie commented on src/aiida_qe_xspec/calculations/functions/xspectra/get_xps_spectra.py:
```diff
@@ -3,109 +3,105 @@
+    equivalency_data = equivalent_sites_data.get_dict()
+
+    data_dict = {element: {} for element in core_levels.keys()}
+    for key, xspectra_out_params in ch_nodes.items():
+        element, _, site, orbital = key.split('_')
+        data_dict[element].setdefault(orbital, {})
+        energy = xspectra_out_params.get_dict()['energy']
+        data_dict[element][orbital][f'site_{site}'] = {'energy': energy, 'multiplicity': equivalency_data[f'site_{site}']['multiplicity']}
```
As this section depends on a specific formatting of the keys for each `ch_node` (e.g. `c_site_1_1s`), *any* deviation from that due to the core level string in `core_levels`, `pseudos`, or `core_level_pseudos` being written incorrectly (e.g. `1_s`, instead of `1s`) needs to be checked before the `XpsWorkChain` is started - otherwise the process will except only at the end of the run.

I would suggest adding input validation for the `XpsWorkChain` to ensure that `key.split()` does actually produce something like `['c', 'site', '1', '1s']` and return an error if not.

superstar54 replied:
Thanks for pointing out this. I have revised the code to use the exact keys (element, orbital, site), instead of using a `string` and `split`., please check the commit [f2d6e3f](https://github.com/aiidaplugins/aiida-qe-xspec/pull/16/commits/f2d6e3fa353e035780e3dc56c14d213b1414b8d6)

PNOGillespie reviewed (changes requested):
Hi @superstar54. Thanks for addressing some of the changes and adding tests. I have one last point which I think maybe wasn't made clear in the previous review.

We can probably go ahead and merge after addressing that last point.

PNOGillespie commented on src/aiida_qe_xspec/workflows/xps.py:
```diff
@@ -25,23 +25,50 @@
-                    f'The ``correction_energies`` provided ({ce_list}) does not match the list of'
-                    f' absorbing elements ({absorbing_elements_list})'
-                )
+    if 'core_levels' in inputs:
+        # keys of core_levels should be a subset of the structure elements and core_hole_pseudos
+        core_levels = inputs['core_levels'].get_dict()
+        if not set(core_levels.keys()).issubset(elements_present):
+            elements_not_present = set(core_levels.keys()) - elements_present
+            raise ValidationError(
+                f'The following elements: {elements_not_present} are not present in the'
+                f' structure ({elements_present}) provided.'
+            )
+        if not set(core_levels.keys()).issubset(set(core_hole_pseudos.keys())):
+            elements_not_present = set(core_levels.keys()) - set(core_hole_pseudos.keys())
+            raise ValidationError(
+                f'The following elements: {elements_not_present} are required for analysis but '
+                f'their pseudopotentials are not provided.'
+            )
```
There needs to be input validation for the `values` of `core_levels` in addition to the `keys`. For instance:

```
core_levels = {'C' : ['1s']}
```
is the correct input format, however:
```
core_levels = {'C' : '1s'}
```
causes the `WorkChain` to except when processing the results, since `get_xps_spectra` is expecting a `list`. Also the core level labels have to be entered in 3 different inputs, which means that any differences in their spelling or formatting will also cause problems.

The previous changes have reduced some potential cases where exceptions could happen at the end, but I think there needs to be explicit validation of the inputs to ensure that `core_levels`, `correction_energies`, and `core_hole_pseudos` use the same labels for the core levels themselves.

I would suggest to add a check to ensure that the core levels given are valid (e.g. must be "1s", "2s", "2p", etc) and have the same format - mainly to check against typos and formatting issues before starting the workflow, since these problems only occur at the end rather than on submission.

superstar54 replied:
Resolved in commit : [f4ea8f3](https://github.com/aiidaplugins/aiida-qe-xspec/pull/16/commits/f4ea8f35cf6d0ce446c88b5ebd6b5b2189448bb0).

PNOGillespie reviewed (changes requested):
Hi @superstar54. Please take a look at the comment for this review - I don't think we're quite there yet with input validation.

Please let me know if anything about it is unclear.

superstar54 commented:
Hi @PNOGillespie, thanks for your feedback!

I've added more input validation along with corresponding tests in this commit: [f4ea8f3](https://github.com/aiidaplugins/aiida-qe-xspec/pull/16/commits/f4ea8f35cf6d0ce446c88b5ebd6b5b2189448bb0).

These improvements should make the input validation more robust and help catch potential issues early. Let me know if you have any further suggestions!

PNOGillespie commented on src/aiida_qe_xspec/workflows/xps.py:
```diff
@@ -93,15 +189,6 @@ def define(cls, spec):
                 ' element. Must use the mapping "{element}" : {Upf}".'
             )
         )
-        spec.input_namespace(
-            'gipaw_pseudos',
-            valid_type=(orm.UpfData, UpfData),
-            dynamic=True,
-            help=(
-                'Dynamic namespace for pairs of ground-state pseudopotentials for each absorbing'
-                ' element. Must use the mapping "{element}" : {Upf}".'
-            )
-        )
```
Since this part has been removed, and the `gipaw_pseudos` has been merged into `core_hole_pseudos`, the `help` string for `core_hole_pseudos` needs to be updated for the new input format.

Doesn't need to be extensive, just something like:
```
'Dynamic namespace for ground-state and core-hole pseudopotentials for each absorbing'
' element. Must use the mapping "{element}" : {"gipaw" : {Upf}, "1s" : {Upf}}".'

superstar54 replied:
Done, in commit: [3b52fcb](https://github.com/aiidaplugins/aiida-qe-xspec/pull/16/commits/3b52fcb9fb1aa239342b5f9bc8858e9375c87d55)

PNOGillespie reviewed (changes requested):
Thanks for the changes @superstar54. The input scheme is now a lot more robust for workflow users.

Just need to update the `help` string for `core_hole_pseudos` and I think we're done here.

PNOGillespie commented:
Good work @superstar54! Go ahead and merge this.

superstar54 commented:
@PNOGillespie  Thank you for the review!

superstar54 commented:
Hi @PNOGillespie , I think it's good to make another release (`v0.1.1`) to include the new functionality. I will do it now.
